### PR TITLE
Change input data structure from named tuple to a class with slots

### PR DESCRIPTION
**Pull request by wiederm**
## Pull Request Summary

This PR addresses the issue #286.

We had two data structures (NNPInput and NNPInputTuple) to pass information from the dataset to the models. The NNPInput class was designed as dataclass and did input validation and casting to dtype/device during training. The dataclass then constructed a NamedTuple that contained all fields of the dataclass, which was then passed as input to the model --- in other words, the model only ever got the NamedTuple as input. To allow working with Jax models, the Named Tuple contained either jax arrays or pytorch tensors depending on passed keywords.

NamedTuples are useful during training, but in MD simulations the immutable datastructure leads to overhead.

### Key changes
 - [x] Change NamedTuple datastructure to a Python class with slots, and remove the double structure dataclass/namedtuple that was used for convenience. 
- [x] change names in NNPInput and Metadata classes from `E` and `F`` to `per_system_energy` and `per_atom_force` 

This PR proposes the following changes:

## Pull Request Checklist
 - [x] Issue(s) raised/addressed and linked
 - [x] Includes appropriate unit test(s)
 - [x] Appropriate docstring(s) added/updated
 - [x] Appropriate .rst doc file(s) added/updated
 - [x] PR is ready for review

### CODE REVIEW

File: modelforge/dataset/dataset.py [hunk 1]
```diff
@@ -102,217 +103,186 @@ class Metadata:
...
+    )
 
-    def to(
+    def __init__(
         self,
-        *,
-        device: Optional[torch.device] = None,
-        dtype: Optional[torch.dtype] = None,
+        atomic_numbers: torch.Tensor,
+        positions: torch.Tensor,
```
Comment (chrisiacovella) [inline modelforge/dataset/dataset.py:167]: This now drops the option for positions to have units.  I think that is probably fine, but we'll probably want to make a helper function that returns an NNPInput instance,  where positions and box vectors require units to be attached.

**Comment by chrisiacovella** [comment]
I think this is a good change; one concern, you've now started to introduce "per_system" instead of "per_molecule", but it's not consistent.  I think that should be changed over fully before merging, otherwise that is going to be very confusing.

**Review by chrisiacovella — CHANGES_REQUESTED** [review]

File: modelforge/dataset/dataset.py (hunk 1 above)
Comment (wiederm) [reply modelforge/dataset/dataset.py:167]: You mean a second implementation with units?

File: modelforge/dataset/dataset.py [hunk 2]
```diff
@@ -102,217 +103,186 @@ class Metadata:
...
-            "NNPInputTuple", [field.name for field in fields(self)]
-        )
-        return NNPInputTuple(
-            *[convert_to_jax(getattr(self, field.name)) for field in fields(self)]
-        )
+    def to_dtype(self, dtype: torch.dtype):
+        """Move all **relevant** tensors to dtype."""
+        self.positions = self.positions.to(dtype)
+        self.box_vectors = self.box_vectors.to(dtype)
+        assert self.box_vectors.shape == (3, 3)
```
[the remark this answers is not in the document]
Comment (wiederm) [reply modelforge/dataset/dataset.py:256]: good catch! will remove!

File: modelforge/potential/aimnet2.py [hunk 3]
```diff
@@ -6,7 +6,7 @@
 
 from modelforge.potential.utils import Dense
 
-from modelforge.dataset.dataset import NNPInput, NNPInputTuple
+from modelforge.dataset.dataset import NNPInput
```
[the remark this answers is not in the document]
Comment (wiederm) [reply modelforge/potential/aimnet2.py:9]: I was thinking about that too. Is `prop.py` the correct module?

File: modelforge/tests/test_models.py [hunk 4]
```diff
@@ -163,8 +165,9 @@ def test_JAX_wrapping(potential_name, single_batch_with_batchsize, prep_temp_dir
         potential_name=potential_name,
         simulation_environment="JAX",
     )
+    from modelforge.jax import convert_NNPInput_to_jax
 
-    nnp_input = batch.nnp_input.as_jax_namedtuple()
+    nnp_input = convert_NNPInput_to_jax(batch.nnp_input)
     out = model(nnp_input)["per_molecule_energy"]
```
[the remark this answers is not in the document]
Comment (wiederm) [reply modelforge/tests/test_models.py:171]: yes, I will move everything to `per_system` now

File: modelforge/train/training.py [hunk 5]
```diff
@@ -218,7 +218,7 @@ def _get_charges(
             "per_molecule_total_charge_predict": total_charge_predict,
```
[the remark this answers is not in the document]
Comment (wiederm) [reply modelforge/train/training.py]: resolved!

**Pull request merged**